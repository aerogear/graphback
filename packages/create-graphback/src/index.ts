#!/usr/bin/env node

import * as yargs from 'yargs';

import { init } from './init/components';

type Params = {
  name?: string;
  templateName?: string;
  templateUrl?: string;
  filter?: string;
};

const command = '$0 <name>';

export const desc = 'Create Graphback project from available templates';

// tslint:disable-next-line: typedef
export const builder = (args: yargs.Argv) => {
  args.positional('name', {
    describe: 'Project name',
    type: 'string'
  });
  args.option('templateName', {
    describe: 'Name of the predefined template',
    type: 'string'
  });
  args.option('templateUrl', {
    describe:
      'GitHub URL of the template. For example (https://github.com/wtrocki/graphql-serve-example#master)',
    type: 'string'
  });
  args.option('filter', {
    describe:
      'A filter option used to limit the number of displayed templates that only matches it. For example "postgres" will only display templates that have a Postgres backed',
    type: 'string'
  });
};

export async function handler({
  name,
  templateName,
  templateUrl,
  filter
}: Params) {
  await init(name, templateName, templateUrl, filter);
}

if (require.main === module) {
  // eslint-disable-next-line no-unused-expressions
  yargs
    .command(command, desc, builder, handler)
    .example(
      '$0 my-awesome-project',
      'creates the "my-awesome-project" in the current working directory'
    )
    .demandCommand(1)
    .strict()
    .recommendCommands()
    .help()
    .alias('h', 'help')
    .version()
    .alias('v', 'version').argv;
}
