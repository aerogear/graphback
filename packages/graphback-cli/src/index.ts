#!/usr/bin/env node

import * as yargs from 'yargs';

if (require.main === module) {
  // eslint-disable-next-line no-unused-expressions
  yargs
    .commandDir('commands')
    .demandCommand(1)
    .strict()
    .recommendCommands()
    .help()
    .alias('h', 'help')
    .version()
    .alias('v', 'version')
    .argv;
}
