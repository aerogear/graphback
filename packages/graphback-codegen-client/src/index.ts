export * from './ClientDocuments'
export * from './ClientCRUDPlugin'