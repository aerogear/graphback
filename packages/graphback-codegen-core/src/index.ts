 
// Top level API
export * from './api'
export * from './graphql'
