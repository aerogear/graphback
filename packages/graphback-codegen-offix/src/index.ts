export * from './OffixPlugin'
export * from './OffixRuntimeData'


//Required for plugins
export { OffixPlugin as Plugin } from './OffixPlugin'