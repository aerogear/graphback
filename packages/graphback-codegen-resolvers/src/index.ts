export * from "./api/OutputResolver"
export * from "./api/generateResolvers"
export * from './api/ResolverGeneratorOptions'