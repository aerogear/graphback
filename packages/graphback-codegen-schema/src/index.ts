export * from './writer/schemaFormatters'
export * from './SchemaCRUDPlugin'