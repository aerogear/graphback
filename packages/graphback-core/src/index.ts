
//Top level API
export * from './crud'
export * from './plugin/GraphbackPlugin'
export * from './plugin/GraphbackPluginEngine'
export * from './plugin/GraphbackGlobalConfig'
export * from './plugin/GraphbackCRUDGeneratorConfig';
export * from './plugin/ModelDefinition'
export * from './plugin/getSelectedFieldsFromResolverInfo';
export * from './plugin/GraphbackCoreMetadata'

export * from './relationships/RelationshipMetadataBuilder';
export * from './relationships/relationshipHelpers';
export * from './annotations/DefaultValueAnnotation';

export * from './utils/printSchemaWithDirectives';
export * from './utils/metadataAnnotations';
export * from './utils/fieldTransformHelpers';
export * from './utils/isTransientField';

export * from './runtime';
export * from './db';

export * from './scalars'
