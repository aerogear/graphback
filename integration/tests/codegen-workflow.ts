/* eslint-disable import/no-extraneous-dependencies */
import { readFileSync, rmdirSync } from 'fs';
import * as path from 'path';
import { ApolloServer, PubSub, gql } from "apollo-server";
import { createTestClient, ApolloServerTestClient } from 'apollo-server-testing';
import { loadConfig } from 'graphql-config';
import { loadResolversFiles } from '@graphql-toolkit/file-loading';
import { loadDocuments } from '@graphql-toolkit/core';
import { GraphQLFileLoader } from '@graphql-toolkit/graphql-file-loader';
import * as Knex from 'knex';
import { migrateDB } from '../../packages/graphql-migrations/src';
import { GraphbackGenerator } from "../../packages/graphback/src";
import { createKnexPGCRUDRuntimeServices } from "../../packages/graphback-runtime-knex/src"

/** global config */
let client: any;
let db: Knex;

beforeAll(async (done) => {
    const { projectConfig, graphbackConfig, dbMigrationsConfig } = await getConfig();

    const modelText = readFileSync(graphbackConfig.model, 'utf8');
    const generator = new GraphbackGenerator(modelText, graphbackConfig);
    generator.generateSourceCode();

    const schemaString = await projectConfig.getSchema('string');
    expect(schemaString).toMatchSnapshot();

    const knex = Knex(dbMigrationsConfig);
    await knex.schema.dropTableIfExists('comment').dropTableIfExists('commentmetadata').dropTableIfExists('note')

    const schema = await projectConfig.getSchema();
    const { newDB } = await migrateDB(dbMigrationsConfig, schema);

    expect(newDB).toMatchSnapshot();

    await seedDatabase(knex);

    const resolvers = loadResolversFiles([
        path.resolve('./output/resolvers/graphbackResolvers.ts'),
        path.resolve('./mocks/resolvers/resolvers.ts')
    ])

    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const { models } = require('../output/resolvers/models');

    const pubSub = new PubSub();
    const context = createKnexPGCRUDRuntimeServices(models, schema, knex, pubSub);

    const server = new ApolloServer({
        typeDefs: await projectConfig.getSchema('DocumentNode'),
        resolvers,
        context
    });

    client = createTestClient(server);
    db = knex;

    done();
})

afterAll(async (done) => {
    rmdirSync(path.resolve('./output'), { recursive: true });
    await db.destroy();
    done();
});

async function seedDatabase(db: Knex) {
    await db('note').insert([
        {
            title: 'Note A',
            description: 'Note A Description'
        },
        {
            title: 'Note B',
            description: 'Note B Description'
        }
    ]);

    await db('commentmetadata').insert([
        {
            opened: true
        },
        {
            opened: false
        }
    ])

    await db('comment').insert([
        {
            text: 'Note A Comment',
            description: 'Note A Comment Description',
            noteId: 1,
            metadataId: 1
        },
        {
            text: 'Note A Comment 2',
            description: 'Note A Comment Description',
            noteId: 1,
            metadataId: 2
        }
    ]);
}

const getConfig = async () => {
    const config = await loadConfig({
        rootDir: process.cwd(),
        extensions: [
            () => ({ name: 'graphback' }),
            () => ({ name: 'dbmigrations' })
        ]
    });

    const projectConfig = config.getDefault();
    const graphbackConfig = projectConfig.extension('graphback');
    const dbMigrationsConfig = projectConfig.extension('dbmigrations');

    return { projectConfig, graphbackConfig, dbMigrationsConfig };
}

/**
 * Helper to load a client document from a file
 * 
 * @param name Client document name
 */
async function getDocument(name: string) {
    const documents = await loadDocuments(path.resolve(`./output/client/**/${name}.graphql`), {
        loaders: [
            new GraphQLFileLoader()
        ]
    });

    return documents[0];
}

test('Find all notes', async (done) => {
    const { document } = await getDocument('findAllNotes');

    const { data } = await client.query({ query: document });

    expect(data).toBeDefined();
    expect(data.findAllNotes).toEqual([
        {
            id: '1',
            title: 'Note A',
            description: 'Note A Description',
            comments: [
                {
                    id: '1',
                    text: 'Note A Comment',
                    description: 'Note A Comment Description'
                },
                {
                    id: '2',
                    text: 'Note A Comment 2',
                    description: 'Note A Comment Description'
                }
            ]
        },
        {
            id: '2',
            title: 'Note B',
            description: 'Note B Description',
            comments: []
        }
    ])

    done();
})

test('Find all notes except the first', async (done) => {
    const { document } = await getDocument('findAllNotes');

    const { data } = await client.query({ query: document, variables: { offset: 1 } });

    expect(data).toBeDefined();
    expect(data.findAllNotes).toEqual([
        {
            id: '2',
            title: 'Note B',
            description: 'Note B Description',
            comments: []
        }
    ])

    done();
})

test('Find at most one note', async (done) => {
    const { document } = await getDocument('findAllNotes');

    const { data } = await client.query({ query: document, variables: { limit: 1 } });

    expect(data).toBeDefined();
    expect(data.findAllNotes).toEqual([
        {
            id: '1',
            title: 'Note A',
            description: 'Note A Description',
            comments: [
                {
                    id: '1',
                    text: 'Note A Comment',
                    description: 'Note A Comment Description'
                },
                {
                    id: '2',
                    text: 'Note A Comment 2',
                    description: 'Note A Comment Description'
                }
            ]
        },
    ])

    done();
})

test('Find all comments', async done => {
    const { document } = await getDocument('findAllComments');

    const { data } = await client.query({ query: document });

    expect(data).toBeDefined();
    expect(data.findAllComments).toEqual([
        {
            id: '1',
            text: 'Note A Comment',
            description: 'Note A Comment Description',
            note: {
                id: '1',
                title: 'Note A',
                description: 'Note A Description'
            },
            metadata: {
                id: '1',
                opened: true
            }
        },
        {
            id: '2',
            text: 'Note A Comment 2',
            description: 'Note A Comment Description',
            note: {
                id: '1',
                title: 'Note A',
                description: 'Note A Description'
            },
            metadata: {
                id: '2',
                opened: false
            }
        }
    ])

    done();
})

test('Note 1 should be defined', async (done) => {
    const response = await findNote('1', client);
    expect(response.data).toBeDefined();
    const notes = response.data.findNotes;
    expect(notes).toEqual([{
        id: '1',
        title: 'Note A',
        description: 'Note A Description',
        comments: [
            {
                id: '1',
                text: 'Note A Comment',
                description: 'Note A Comment Description'
            },
            {
                id: '2',
                text: 'Note A Comment 2',
                description: 'Note A Comment Description'
            }
        ]
    }]);

    done();
})

test('Note 1 Comments exists', async (done) => {
    const response = await findNoteComments('1', client);
    expect(response.data).toBeDefined()
    expect(response.data.findComments).toHaveLength(2);

    done();
})

test('Find at most one comment on Note 1', async (done) => {

    const { document } = await getDocument('findComments');

    const response = await client.query({ query: document, variables: { fields: { noteId: 1 }, limit: 1 } });
    
    expect(response.data).toBeDefined()
    const notes = response.data.findComments
    expect(notes).toHaveLength(1);
    expect(notes).toEqual([
        {
            id: '1',
            text: 'Note A Comment',
            description: 'Note A Comment Description',
            metadata: {
                id: "1",
                opened: true,
            },
            note: {
                description: "Note A Description",
                id: "1",
                title: "Note A",
            },
        }
    ])

    done();
})

test('Find comments on Note 1 except first', async (done) => {

    const { document } = await getDocument('findComments');

    const response = await client.query({ query: document, variables: { fields: { noteId: 1 }, offset: 1 } });
    
    expect(response.data).toBeDefined()
    const notes = response.data.findComments
    expect(notes).toHaveLength(1);
    expect(notes).toEqual([
        {
            id: '2',
            text: 'Note A Comment 2',
            description: 'Note A Comment Description',
            metadata: {
                "id": "2",
                "opened": false,
            },
            note: {
                "description": "Note A Description",
                "id": "1",
                "title": "Note A",
            },
        }
    ])

    done();
})

test('Should update Note 1 title', async (done) => {
    const response = await updateNote({ id: '1', title: 'Note 1 New Title' }, client);
    expect(response.data).toBeDefined();
    expect(response.data.updateNote.title).toBe('Note 1 New Title');

    done();
});

test('Should create a new Note', async (done) => {
    const response = await createNote(client, { title: 'New note', description: 'New note description' });
    expect(response.data).toBeDefined();
    expect(response.data.createNote).toEqual({ id: '3', title: 'New note', description: 'New note description' });

    done();
})

test('Delete Note 1', async done => {
    const response = await deleteNote(client, { id: '2' });
    expect(response.data).toBeDefined();
    expect(response.data.deleteNote).toEqual({ id: '2', description: 'Note B Description', title: 'Note B' });

    done();
});

test('Test custom query', async done => {
    const document = gql`
        query {
            helloWorld
        }
    `;
    const response = await client.query({ query: document })

    expect(response.data).toEqual({ helloWorld: 'Hello!' });

    done();
})

async function updateNote(input: any, client: ApolloServerTestClient) {
    const { document } = await getDocument('updateNote');

    const response = await client.mutate({ mutation: document, variables: { input } });

    return response;
}

async function createNote(client: ApolloServerTestClient, input: any) {
    const { document } = await getDocument('createNote');

    const response = await client.mutate({ mutation: document, variables: { input } });

    return response;
}

async function deleteNote(client: ApolloServerTestClient, input: any) {
    const { document } = await getDocument('deleteNote');

    const response = await client.mutate({ mutation: document, variables: { input } });

    return response;
}

async function findNote(id: string, client: ApolloServerTestClient) {
    const { document } = await getDocument('findNotes');

    const response = await client.query({ query: document, variables: { fields: { id } } });

    return response;
}

async function findNoteComments(noteId: string, client: ApolloServerTestClient) {
    const { document } = await getDocument('findComments');

    const response = await client.query({ query: document, variables: { fields: { noteId } } });

    return response;
}

