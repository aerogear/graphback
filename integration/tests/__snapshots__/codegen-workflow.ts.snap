// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"\\"\\"\\" @model \\"\\"\\"
type Comment {
  id: ID!
  text: String
  description: String
  \\"\\"\\"@oneToOne key: 'metadataId'\\"\\"\\"
  metadata: CommentMetadata
  \\"\\"\\"@manyToOne field: 'comments', key: 'noteId'\\"\\"\\"
  note: Note
}

input CommentInput {
  id: ID
  text: String
  description: String
  metadataId: ID
  noteId: ID
}

\\"\\"\\"@model\\"\\"\\"
type CommentMetadata {
  id: ID!
  opened: Boolean
}

input CommentMetadataInput {
  id: ID
  opened: Boolean
}

type Mutation {
  createNote(input: NoteInput): Note!
  updateNote(input: NoteInput): Note!
  deleteNote(input: NoteInput): Note!
  createComment(input: CommentInput): Comment!
  updateComment(input: CommentInput): Comment!
  deleteComment(input: CommentInput): Comment!
  createCommentMetadata(input: CommentMetadataInput): CommentMetadata!
  updateCommentMetadata(input: CommentMetadataInput): CommentMetadata!
  deleteCommentMetadata(input: CommentMetadataInput): CommentMetadata!
}

\\"\\"\\" @model \\"\\"\\"
type Note {
  id: ID!
  title: String!
  description: String
  \\"\\"\\"@oneToMany field: 'note', key: 'noteId'\\"\\"\\"
  comments: [Comment]!
}

input NoteInput {
  id: ID
  title: String
  description: String
}

type Query {
  helloWorld: String
  findAllNotes(limit: Int, offset: Int): [Note]!
  findNotes(fields: NoteInput, limit: Int, offset: Int): [Note]!
  findAllComments(limit: Int, offset: Int): [Comment]!
  findComments(fields: CommentInput, limit: Int, offset: Int): [Comment]!
  findAllCommentMetadata(limit: Int, offset: Int): [CommentMetadata]!
  findCommentMetadata(fields: CommentMetadataInput, limit: Int, offset: Int): [CommentMetadata]!
}

type Subscription {
  newNote(input: NoteInput): Note!
  updatedNote(input: NoteInput): Note!
  deletedNote(input: NoteInput): Note!
  newComment(input: CommentInput): Comment!
  updatedComment(input: CommentInput): Comment!
  deletedComment(input: CommentInput): Comment!
  newCommentMetadata(input: CommentMetadataInput): CommentMetadata!
  updatedCommentMetadata(input: CommentMetadataInput): CommentMetadata!
  deletedCommentMetadata(input: CommentMetadataInput): CommentMetadata!
}

schema {
  query: Query
  mutation: Mutation
  subscription: Subscription
}
"
`;

exports[` 2`] = `
Object {
  "tableMap": Map {
    "Note" => Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "title" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "title",
          "nullable": false,
          "type": "string",
        },
        "description" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "title",
          "nullable": false,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "note",
      "primaries": Array [
        Object {
          "columns": Array [
            "id",
          ],
          "name": "note_pkey",
        },
      ],
      "uniques": Array [],
    },
    "Comment" => Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "text" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "text",
          "nullable": true,
          "type": "string",
        },
        "description" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
        "metadata" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "commentmetadata",
            "type": "CommentMetadata",
          },
          "name": "metadataId",
          "nullable": true,
          "type": "integer",
        },
        "note" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "note",
            "type": "Note",
          },
          "name": "noteId",
          "nullable": true,
          "type": "integer",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "text",
          "nullable": true,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "commentmetadata",
            "type": "CommentMetadata",
          },
          "name": "metadataId",
          "nullable": true,
          "type": "integer",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "note",
            "type": "Note",
          },
          "name": "noteId",
          "nullable": true,
          "type": "integer",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "comment",
      "primaries": Array [
        Object {
          "columns": Array [
            "id",
          ],
          "name": "comment_pkey",
        },
      ],
      "uniques": Array [],
    },
    "CommentMetadata" => Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "opened" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "opened",
          "nullable": true,
          "type": "boolean",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "opened",
          "nullable": true,
          "type": "boolean",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "commentmetadata",
      "primaries": Array [
        Object {
          "columns": Array [
            "id",
          ],
          "name": "commentmetadata_pkey",
        },
      ],
      "uniques": Array [],
    },
  },
  "tables": Array [
    Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "title" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "title",
          "nullable": false,
          "type": "string",
        },
        "description" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "title",
          "nullable": false,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "note",
      "primaries": Array [
        Object {
          "columns": Array [
            "id",
          ],
          "name": "note_pkey",
        },
      ],
      "uniques": Array [],
    },
    Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "text" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "text",
          "nullable": true,
          "type": "string",
        },
        "description" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
        "metadata" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "commentmetadata",
            "type": "CommentMetadata",
          },
          "name": "metadataId",
          "nullable": true,
          "type": "integer",
        },
        "note" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "note",
            "type": "Note",
          },
          "name": "noteId",
          "nullable": true,
          "type": "integer",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "text",
          "nullable": true,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "commentmetadata",
            "type": "CommentMetadata",
          },
          "name": "metadataId",
          "nullable": true,
          "type": "integer",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "note",
            "type": "Note",
          },
          "name": "noteId",
          "nullable": true,
          "type": "integer",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "comment",
      "primaries": Array [
        Object {
          "columns": Array [
            "id",
          ],
          "name": "comment_pkey",
        },
      ],
      "uniques": Array [],
    },
    Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "opened" => Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "opened",
          "nullable": true,
          "type": "boolean",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "name": "opened",
          "nullable": true,
          "type": "boolean",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "commentmetadata",
      "primaries": Array [
        Object {
          "columns": Array [
            "id",
          ],
          "name": "commentmetadata_pkey",
        },
      ],
      "uniques": Array [],
    },
  ],
}
`;
